tests/unit_tests/sparse_blas/include/test_common.hpp

```diff
@@ -0,0 +1,248 @@
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#ifndef _TEST_COMMON_HPP__
+#define _TEST_COMMON_HPP__
+
+#include <iostream>
+#include <memory>
+#include <limits>
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+#include "test_helper.hpp"
+
+// Sparse BLAS domain needs to call more functions per test so we use this macro helper to select between runtime and compile dispatch for each function
+#ifdef CALL_RT_API
+#define CALL_RT_OR_CT(FUNC, QUEUE, ...) FUNC(QUEUE, __VA_ARGS__)
+#else
+#define CALL_RT_OR_CT(FUNC, QUEUE, ...) TEST_RUN_CT_SELECT(QUEUE, FUNC, __VA_ARGS__);
+#endif
+
+template <typename T>
+struct complex_info {
+    using real_type = T;
+    static const bool is_complex = false;
+};
+
+template <typename T>
+struct complex_info<std::complex<T>> {
+    using real_type = T;
+    static const bool is_complex = true;
+};
+
+void print_error_code(sycl::exception const &e);
+
+// Catch asynchronous exceptions.
+struct exception_handler_t {
+    void operator()(sycl::exception_list exceptions) {
+        for (std::exception_ptr const &e : exceptions) {
+            try {
+                std::rethrow_exception(e);
+            }
+            catch (sycl::exception const &e) {
+                std::cout << "Caught asynchronous SYCL exception:\n" << e.what() << std::endl;
+                print_error_code(e);
+            }
+        }
+    }
+};
+
+// Use a unique_ptr to automatically free device memory on unique_ptr destruction.
+template <class T>
+auto malloc_device_uptr(sycl::queue &q, std::size_t num_elts) {
+    struct Deleter {
+        sycl::queue &q;
+        Deleter(sycl::queue &_q) : q(_q) {}
+        void operator()(T *ptr) {
+            sycl::free(ptr, q);
+        }
+    };
+    return std::unique_ptr<T, Deleter>(sycl::malloc_device<T>(num_elts, q), Deleter(q));
+}
+
+// SYCL buffer creation helper.
+template <typename vec>
+sycl::buffer<typename vec::value_type, 1> make_buffer(const vec &v) {
+    sycl::buffer<typename vec::value_type, 1> buf(v.data(), sycl::range<1>(v.size()));
+    return buf;
+}
+
+template <typename fpType>
+struct rand_scalar {
+    inline fpType operator()(double min, double max) {
+        return (fpType(std::rand()) / fpType(RAND_MAX)) * fpType(max - min) - fpType(min);
+    }
+};
+
+template <typename fpType>
+struct rand_scalar<std::complex<fpType>> {
+    inline std::complex<fpType> operator()(double min, double max) {
+        rand_scalar<fpType> rand;
+        return std::complex<fpType>(rand(min, max), rand(min, max));
+    }
+};
+
+template <typename fpType>
+void rand_vector(std::vector<fpType> &v, std::size_t n) {
+    using fpRealType = typename complex_info<fpType>::real_type;
+    v.resize(n);
+    rand_scalar<fpType> rand;
+    for (std::size_t i = 0; i < n; i++) {
+        v[i] = rand(fpRealType(-0.5), fpRealType(0.5));
```

Comment:
I may be completely wrong here, but could you please print and verify that you are getting non-zero imaginary parts in `complex` random vectors generated through this function?

My understanding is that for `fpType = std::complex<float>`, the compiler would see 2 options for creating `rand_scalar<fpType> rand`: `template <typename std::complex<float>> struct rand_scalar`, and (here's the problem, I believe:) `template <typename std::complex<float>> rand_scalar<std::complex<std::complex<float>>>`.

I believe that `template <typename float> struct rand_scalar<std::complex<float>>` option is never generated/used because you are explicitly stating the template to be used instead of the compiler deducing it, and these lines actually end up using the `rand_scalar` meant for reals instead of complex, which means `v[i]` contains only real part and not the imaginary part.

Reply:
This is working as expected. The first test using complex prints:
```
(0.957702,0.563096)
(0.73828,1.47063)
(1.40221,1.35092)
(0.766666,1.03976)
(0.875207,1.26025)
(1.01254,1.16772)
```
for `x_host`.

On line 110 the compiler is forced to instantiate an object of type `rand_scalar<std::complex<float>>` when `rand_vector` is called with a vector of `std::vector<std::complex<float>>`.
Even if there was a confusion the `operator()` would return a `std::complex<std::complex<float>>` value which cannot be casted to `std::complex<float>` when assigning it to `v[i]`.